src/Orleans.Transactions/InClusterTM/TransactionManager.cs

```diff
@@ -5,10 +5,99 @@
+        private PeriodicAction monitor;
+        public TransactionManagerMetrics(ITelemetryProducer telemetryProducer, TimeSpan metricReportInterval)
+        {
+            this.telemetryProducer = telemetryProducer;
+            this.monitor = new PeriodicAction(metricReportInterval, this.ReportMetrics);
+        }
+
+        public void TryReportMetrics()
+        {
+            this.monitor.TryAction(DateTime.Now);
+        }
+
+        private void ResetCounters(DateTime lastReportTime)
+        {
+            this.lastReportTime = lastReportTime;
+            this.StartTransactionRequestCounter = 0;
+            this.AbortTransactionRequestCounter = 0;
+            this.CommitTransactionRequestCounter = 0;
+
+            this.AbortedTransactionDueToDependencyCounter = 0;
+            this.AbortedTransactionDueToMissingInfoInTransactionTableCounter = 0;
+            this.PendingTransactionCounter = 0;
+            this.ValidatedTransactionCounter = 0;
+            this.AbortedTransactionCounter = 0;
+        }
+
+        private void ReportMetrics()
+        {
+            if (this.telemetryProducer == null)
+                return;
+            var now = DateTime.Now;
+            var timeSinceLastReportInSeconds = Math.Max(1, (now - this.lastReportTime).TotalSeconds);
+            var startTransactionTps = StartTransactionRequestCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(StartTransactionRequestTPS, startTransactionTps);
+
+            var abortTransactionTPS = AbortTransactionRequestCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(AbortTransactionRequestTPS, abortTransactionTPS);
+
+            var commitTransactionTPS = CommitTransactionRequestCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(CommitTransactionRequestTPS, commitTransactionTPS);
+
+            var abortedTransactionDueToDependentTPS = AbortedTransactionDueToDependencyCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(AbortedTransactionDueToDependencyTPS, abortedTransactionDueToDependentTPS);
+
+            var abortedTransactionDueToMissingInfoInTransactionTableTPS = AbortedTransactionDueToMissingInfoInTransactionTableCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(AbortedTransactionDueToMissingInfoInTransactionTableTPS, abortedTransactionDueToMissingInfoInTransactionTableTPS);
+
+            var pendingTransactionTPS = PendingTransactionCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(PendingTransactionTPS, pendingTransactionTPS);
+
+            var validatedTransactionTPS = ValidatedTransactionCounter / timeSinceLastReportInSeconds;
+            this.telemetryProducer.TrackMetric(ValidatedTransactionTPS, validatedTransactionTPS);
+
+            var abortedTransactionTPS = AbortedTransactionCounter / timeSinceLastReportInSeconds;
```

Comment:
To do this accurately (I'm not sure if that's a concern for us), we need to read the field using `Interlocked.Exchange` to simultaneously read the value and reset it to 0

Reply:
this class is only used in grain based TM, so it is single threaded. I assume I don't need to handle threading issue there.

Reply:
Is this only used in grain based TM because the external TM isn't merged? I think we have only one TransactionManager class which is wrapped either in a grain or not depending on where it's being hosted. Is that right?

Reply:
There's only one grain based TM in Orleans master I think. The external TM is not merged in, and I think it may/will not be in the future either AFAIK. 

Yes it is only used in the grain based TM.

Reply:
There is only one grain-based TM, but this is not a grain-based TM. This component is shared between both implementations - that's why all of the collections are concurrent collections

Reply:
I meant, there's only one TM, which is the grain based one. And this component is only used the the grain based one. If it will be used in another TM which is not single threaded in the future, let's take care of that in the future. But I don't see why we should take that unknown future into account now. 

All the collection are concurrent collection probably because of the old external TM. But that TM is not merged in, and I doubt it will ever be. Hence my conclusion on it is safe to not use Interlocked.Exchange in this component.

Reply:
Ok, no problem - as long as @jason-bragg is ok with it. We can eventually migrate to non-concurrent collections and improve perf (a large part of the memory overhead was in ConcurrentDictionary/ConcurrentQueue maintenance)

Reply:
@jason-bragg this maybe lost in diff updates. But here is the place we need your confirmation too

Reply:
There is no plan at this time to pull in the external TM.  The next major architectural change will be to move to a distributed TM, not an external TM.  As we clean up this TM, the design decisions made due to multi-threaded access will be removed.  Interlocked.Exchange should not be necessary.